### ISSUE TITLE
broken with molecule HEAD

### ISSUE THREAD

**Issue by apatard** [ansible-community/molecule-vagrant#43]
Since commit https://github.com/ansible-community/molecule/commit/65ccd6beab63ed3c4901c9ef6296461fe8f375e5, every time a ``molecule.config.Config`` object is created, a lock is taken, to avoid concurrent molecule run.

The molecule-vagrant code is using ``molecule.config.Config`` inside the ``vagrant.py`` module, in order to get information like ephemeral directory, path to the ``Vagrantfile`` to write. So, during the destroy / converge / create steps, the module will now try to get the lock already taken by molecule, making things fail.

I'm not sure it's a molecule or molecule-vagrant bug. A possible fix would look like : https://gist.github.com/apatard/b037eb8313201399a33df5677bbaae64 but it's annoying to have to compute things inside the ``vagrant.py`` file, which were available through molecule before.

Let's make the issue visible and discuss it

**Comment by konstruktoid** [comment]
Is this the situation you're referring to?

```
    failed: [localhost] (item=bionic) => {"ansible_loop_var": "item", "changed": false, "cmd": ["/usr/local/bin/vagrant", "status", "--machine-readable", "bionic"], "item": {"box": "bento/ubuntu-18.04", "instance_raw_config_args":
 ["vbguest.installer_options = { allow_kernel_upgrade: true }"], "memory": 1024, "name": "bionic", "provider_options": {"nic_model_type": "\"Am79C973\"", "vm.boot_timeout": 600}}, "msg": "ERROR: See log file '/Users/xyzx/.cache/mo
lecule/ansible-role-hardening/default/vagrant-bionic.err'", "rc": 1, "stderr": "### 2020-09-29 18:02:45 ###\n### 2020-09-29 18:02:45 ###\n### 2020-09-29 18:02:45 ###\n### 2020-09-29 18:36:51 ###\n### 2020-09-29 18:36:51 ###\n### 2
020-09-29 18:36:51 ###\n### 2020-09-29 18:57:00 ###\n### 2020-09-29 18:58:02 ###\n### 2020-09-29 19:00:46 ###\n### 2020-09-29 19:00:46 ###\n### 2020-09-29 19:00:46 ###\n### 2020-09-29 19:34:05 ###\n### 2020-09-29 19:34:05 ###\n###
 2020-09-29 19:34:05 ###\n### 2020-09-29 19:50:47 ###\n### 2020-09-29 21:47:27 ###\n### 2020-09-30 08:17:01 ###\n### 2020-09-30 08:17:01 ###\n### 2020-09-30 08:17:01 ###\n### 2020-09-30 08:41:55 ###\n### 2020-09-30 08:41:55 ###\n#
## 2020-09-30 08:41:55 ###\n### 2020-09-30 09:20:41 ###\n### 2020-09-30 09:20:41 ###\n### 2020-09-30 09:20:41 ###\n### 2020-09-30 09:38:24 ###\n### 2020-09-30 09:38:52 ###\n### 2020-09-30 09:38:52 ###\n### 2020-09-30 09:38:52 ###\
n### 2020-09-30 09:58:36 ###\n### 2020-09-30 09:58:36 ###\n### 2020-09-30 09:58:36 ###\n### 2020-09-30 10:01:33 ###\n### 2020-09-30 10:02:28 ###\n### 2020-09-30 10:05:00 ###\n### 2020-09-30 10:05:31 ###\n### 2020-09-30 10:05:31 ##
#\n### 2020-09-30 10:05:31 ###\n### 2020-09-30 10:20:15 ###\n### 2020-09-30 10:20:15 ###\n### 2020-09-30 10:20:15 ###\n### 2020-09-30 10:26:11 ###\n### 2020-09-30 10:26:33 ###\n### 2020-09-30 10:26:33 ###\n### 2020-09-30 10:26:33
###\n### 2020-09-30 10:59:41 ###\n### 2020-09-30 10:59:41 ###\n### 2020-09-30 10:59:41 ###\n### 2020-09-30 12:02:07 ###\n### 2020-09-30 12:02:07 ###\n### 2020-09-30 12:03:39 ###\n### 2020-09-30 12:03:39 ###\nAn action 'halt' was a
ttempted on the machine 'bionic',\nbut another process is already executing an action on the machine.\nVagrant locks each machine for access by only one process at a time.\nPlease wait until the other Vagrant process finishes modi
fying this\nmachine, then try again.\n\nIf you believe this message is in error, please check the process\nlisting for any \"ruby\" or \"vagrant\" processes and kill them. Then\ntry again.\n### 2020-09-30 12:03:39 ###\nAn action '
destroy' was attempted on the machine 'bionic',\nbut another process is already executing an action on the machine.\nVagrant locks each machine for access by only one process at a time.\nPlease wait until the other Vagrant process
 finishes modifying this\nmachine, then try again.\n\nIf you believe this message is in error, please check the process\nlisting for any \"ruby\" or \"vagrant\" processes and kill them. Then\ntry again.\n### 2020-09-30 12:04:43 ##
#\n### 2020-09-30 12:05:06 ###\nThere was an error while executing `VBoxManage`, a CLI used by Vagrant\nfor controlling VirtualBox. The command and stderr is shown below.\n\nCommand: [\"showvminfo\", \"7a633308-8e4b-4f0b-add7-7823
c5b4605e\", \"--machinereadable\"]\n\nStderr: VBoxManage: error: Could not find a registered machine with UUID {7a633308-8e4b-4f0b-add7-7823c5b4605e}\nVBoxManage: error: Details: code VBOX_E_OBJECT_NOT_FOUND (0x80bb0001), componen
t VirtualBoxWrap, interface IVirtualBox, callee nsISupports\nVBoxManage: error: Context: \"FindMachine(Bstr(VMNameOrUuid).raw(), machine.asOutParam())\" at line 2715 of file VBoxManageInfo.cpp\n### 2020-09-30 12:05:06 ###\n", "std
err_lines": ["### 2020-09-29 18:02:45 ###", "### 2020-09-29 18:02:45 ###", "### 2020-09-29 18:02:45 ###", "### 2020-09-29 18:36:51 ###", "### 2020-09-29 18:36:51 ###", "### 2020-09-29 18:36:51 ###", "### 2020-09-29 18:57:00 ###",
"### 2020-09-29 18:58:02 ###", "### 2020-09-29 19:00:46 ###", "### 2020-09-29 19:00:46 ###", "### 2020-09-29 19:00:46 ###", "### 2020-09-29 19:34:05 ###", "### 2020-09-29 19:34:05 ###", "### 2020-09-29 19:34:05 ###", "### 2020-09-
29 19:50:47 ###", "### 2020-09-29 21:47:27 ###", "### 2020-09-30 08:17:01 ###", "### 2020-09-30 08:17:01 ###", "### 2020-09-30 08:17:01 ###", "### 2020-09-30 08:41:55 ###", "### 2020-09-30 08:41:55 ###", "### 2020-09-30 08:41:55 #
##", "### 2020-09-30 09:20:41 ###", "### 2020-09-30 09:20:41 ###", "### 2020-09-30 09:20:41 ###", "### 2020-09-30 09:38:24 ###", "### 2020-09-30 09:38:52 ###", "### 2020-09-30 09:38:52 ###", "### 2020-09-30 09:38:52 ###", "### 202
0-09-30 09:58:36 ###", "### 2020-09-30 09:58:36 ###", "### 2020-09-30 09:58:36 ###", "### 2020-09-30 10:01:33 ###", "### 2020-09-30 10:02:28 ###", "### 2020-09-30 10:05:00 ###", "### 2020-09-30 10:05:31 ###", "### 2020-09-30 10:05
:31 ###", "### 2020-09-30 10:05:31 ###", "### 2020-09-30 10:20:15 ###", "### 2020-09-30 10:20:15 ###", "### 2020-09-30 10:20:15 ###", "### 2020-09-30 10:26:11 ###", "### 2020-09-30 10:26:33 ###", "### 2020-09-30 10:26:33 ###", "##
# 2020-09-30 10:26:33 ###", "### 2020-09-30 10:59:41 ###", "### 2020-09-30 10:59:41 ###", "### 2020-09-30 10:59:41 ###", "### 2020-09-30 12:02:07 ###", "### 2020-09-30 12:02:07 ###", "### 2020-09-30 12:03:39 ###", "### 2020-09-30
12:03:39 ###", "An action 'halt' was attempted on the machine 'bionic',", "but another process is already executing an action on the machine.", "Vagrant locks each machine for access by only one process at a time.", "Please wait u
ntil the other Vagrant process finishes modifying this", "machine, then try again.", "", "If you believe this message is in error, please check the process", "listing for any \"ruby\" or \"vagrant\" processes and kill them. Then",
 "try again.", "### 2020-09-30 12:03:39 ###", "An action 'destroy' was attempted on the machine 'bionic',", "but another process is already executing an action on the machine.", "Vagrant locks each machine for access by only one p
rocess at a time.", "Please wait until the other Vagrant process finishes modifying this", "machine, then try again.", "", "If you believe this message is in error, please check the process", "listing for any \"ruby\" or \"vagrant
\" processes and kill them. Then", "try again.", "### 2020-09-30 12:04:43 ###", "### 2020-09-30 12:05:06 ###", "There was an error while executing `VBoxManage`, a CLI used by Vagrant", "for controlling VirtualBox. The command and
stderr is shown below.", "", "Command: [\"showvminfo\", \"7a633308-8e4b-4f0b-add7-7823c5b4605e\", \"--machinereadable\"]", "", "Stderr: VBoxManage: error: Could not find a registered machine with UUID {7a633308-8e4b-4f0b-add7-7823
c5b4605e}", "VBoxManage: error: Details: code VBOX_E_OBJECT_NOT_FOUND (0x80bb0001), component VirtualBoxWrap, interface IVirtualBox, callee nsISupports", "VBoxManage: error: Context: \"FindMachine(Bstr(VMNameOrUuid).raw(), machine
.asOutParam())\" at line 2715 of file VBoxManageInfo.cpp", "### 2020-09-30 12:05:06 ###"]}
    An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'AnsibleModule' object has no attribute 'result'
    failed: [localhost] (item=buster) => {"ansible_loop_var": "item", "changed": false, "item": {"box": "bento/debian-10", "instance_raw_config_args": ["vbguest.installer_options = { allow_kernel_upgrade: true }"], "memory": 1024, "name": "buster"}, "module_stderr": "Traceback (most recent call last):\n  File \"/Users/xyzx/.ansible/tmp/ansible-tmp-1601460426.373004-219958772466214/AnsiballZ_vagrant.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/Users/xyzx/.ansible/tmp/ansible-tmp-1601460426.373004-219958772466214/AnsiballZ_vagrant.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/Users/xyzx/.ansible/tmp/ansible-tmp-1601460426.373004-219958772466214/AnsiballZ_vagrant.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/imp.py\", line 234, in load_module\n    return load_source(name, filename, file)\n  File \"/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/imp.py\", line 170, in load_source\n    module = _exec(spec, sys.modules[name])\n  File \"<frozen importlib._bootstrap>\", line 608, in _exec\n  File \"<frozen importlib._bootstrap_external>\", line 678, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 205, in _call_with_frames_removed\n  File \"/var/folders/g5/cz5cgtwd5l59bzkjkr0cbj0xwnlwnj/T/ansible_vagrant_payload_awoa8wbq/__main__.py\", line 632, in <module>\n  File \"/var/folders/g5/cz5cgtwd5l59bzkjkr0cbj0xwnlwnj/T/ansible_vagrant_payload_awoa8wbq/__main__.py\", line 628, in main\nAttributeError: 'AnsibleModule' object has no attribute 'result'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
    An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'AnsibleModule' object has no attribute 'result'
    failed: [localhost] (item=centos) => {"ansible_loop_var": "item", "changed": false, "item": {"box": "bento/centos-8", "instance_raw_config_args": ["vbguest.installer_options = { allow_kernel_upgrade: true }"], "memory": 1024, "name": "centos", "provider_options": {"nic_model_type": "\"82543GC\"", "vm.boot_timeout": 600}}, "module_stderr": "Traceback (most recent call last):\n  File \"/Users/xyzx/.ansible/tmp/ansible-tmp-1601460431.442317-210655325142642/AnsiballZ_vagrant.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/Users/xyzx/.ansible/tmp/ansible-tmp-1601460431.442317-210655325142642/AnsiballZ_vagrant.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/Users/xyzx/.ansible/tmp/ansible-tmp-1601460431.442317-210655325142642/AnsiballZ_vagrant.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/imp.py\", line 234, in load_module\n    return load_source(name, filename, file)\n  File \"/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/imp.py\", line 170, in load_source\n    module = _exec(spec, sys.modules[name])\n  File \"<frozen importlib._bootstrap>\", line 608, in _exec\n  File \"<frozen importlib._bootstrap_external>\", line 678, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 205, in _call_with_frames_removed\n  File \"/var/folders/g5/cz5cgtwd5l59bzkjkr0cbj0xwnlwnj/T/ansible_vagrant_payload_ttoc7vvk/__main__.py\", line 632, in <module>\n  File \"/var/folders/g5/cz5cgtwd5l59bzkjkr0cbj0xwnlwnj/T/ansible_vagrant_payload_ttoc7vvk/__main__.py\", line 628, in main\nAttributeError: 'AnsibleModule' object has no attribute 'result'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

**Comment by apatard** [comment]
this bug is about failures like:
```
Retrying to acquire lock on /home/zuul/.cache/molecule/scenarios/provider_config_options, waiting for 30 seconds
...
Timedout trying to acquire lock on /home/zuul/.cache/molecule/scenarios/provider_config_options
```
your log seems hard to read but it looks different than the issue here.

**Comment by konstruktoid** [comment]
```
### 2020-09-30 12:03:39 ###\n
An action 'destroy' was attempted on the machine 'bionic',\n
but another process is already executing an action on the machine.\n
Vagrant locks each machine for access by only one process at a time.\n
Please wait until the other Vagrant process
 finishes modifying this\n
machine, then try again.\n\n
If you believe this message is in error, please check the process\n
listing for any \"ruby\" or \"vagrant\" processes and kill them. Then\n
try again.\n
```

**Comment by konstruktoid** [comment]
> this bug is about failures like:
> 
> ```
> Retrying to acquire lock on /home/zuul/.cache/molecule/scenarios/provider_config_options, waiting for 30 seconds
> ...
> Timedout trying to acquire lock on /home/zuul/.cache/molecule/scenarios/provider_config_options
> ```
> 
> your log seems hard to read but it looks different than the issue here.

Yeah, that log was a mess, sorry about that.
The `zuul` issue only arises when a Vagrant machine gets locked and as a result times out.

**Comment by apatard** [comment]
the issue in zuul CI comes from a lock taken by molecule. Your error is talking of a lock taken by vagrant, so we're talking of two different locks. This bug is about the molecule one. so please open a different bug report (with a way to reproduce it if possible).

Moreover, you've not specified your molecule version. I mean this problematic commit is not in any released version of molecule, so unless you're using a clone of molecule, the chances to have the bug mentionned here are very low.

**Comment by konstruktoid** [comment]
Thanks for clarifying @apatard, since no logs where present in this issue and I've had problems with locking (and interpreted the logs (incorrectly) as similar to this) I thought it was related.

Just fyi:
```
$ molecule --version
molecule 3.0.8
   ansible==2.8.0 python==3.6
```

**Comment by sio** [comment]
FYI, it's not just HEAD anymore. After Molecule 3.1 release this lock issue affects the version published at PyPI: https://cirrus-ci.com/task/5285044556136448?command=test#L258

**Comment by ssbarnea** [comment]
Once someone proposes a fix I will help with a release. Sadly I already reached my limit and cannot work directly on it.

That lock is a semaphore for preventing two molecule processes from running the same scenario when run in parallel. It is not supposed to remain open as the lock is released when process is closed (file is not removed).

**Comment by apatard** [comment]
Oh, I thought I had more time before 3.1. Will cook a patch asap

**Comment by OctavioCore** [comment]
I've ran onto an issue that resembles this - after multiple issues with Molecule + Ansible & Molecule-vagrant instalation, i've finally manage to get them up, or so i thought...
When trying to run test sequence, or just anything interacting with Vagrant, Molecule just hangs executing:
`<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/(my username)/.ansible/tmp/ansible-tmp-1602839107.9695609-12707-123778950972138/AnsiballZ_vagrant.py && sleep 0'`
I've checked both Ansible & Vagrant, and they do work, everything else looks fine - normally the step would create machines in VB, but nothing happens. 

After like 5? minutes of waiting, error pops out:
`"module_stdout": "Retrying to acquire lock on /home/(my username)/.cache/molecule/deb.apt-install/default, waiting for 30 seconds` then 60 secs, then 90, then 120... And finally:
`\nTimedout trying to acquire lock on /home/(my username)/.cache/molecule/deb.apt-install/default\n"`

Could this be related/the same issue, or should i open another ticket?

**Comment by apatard** [comment]
> After like 5? Minutes of waiting, error pops out:
> `"module_stdout": "Retrying to acquire lock on /home/(my username)/.cache/molecule/deb.apt-install/default, waiting for 30 seconds` then 60 secs, then 90, then 120... And finally:
> `\nTimedout trying to acquire lock on /home/(my username)/.cache/molecule/deb.apt-install/default\n"`

This message is the symptom of this bug, so no need for an other bug. There'll be a PR for this bug today, as I'm currently running the testsuite with a possible fix

**Comment by apatard** [comment]
Fix pushed as PR 58. Now, let's wait for CI and review.
